views/js/runner/proxy/offline/proxy.js

```diff
@@ -0,0 +1,482 @@
+
+            // we keep items here
+            this.itemStore = itemStoreFactory({
+                preload : true,
+                testId  : config.serviceCallId,
+            });
+
+            this.offlineNavigator = offlineNavigatorFactory(this.itemStore);
+
+            // where we keep actions
+            this.actionStore = null;
+
+            // configuration params, that comes on every request/params
+            this.requestConfig = {};
+
+            // scheduled action promises which supposed to be resolved after action synchronization.
+            this.actionPromises = {};
+
+            // let's you update test data (testData, testContext and testMap)
+            this.dataUpdater = dataUpdater(this.getDataHolder());
+
+            /**
+             * Check whether we have the item in the store
+             *
+             * @param {String} itemIdentifier - the item identifier
+             * @returns {Boolean}
+             */
+            this.hasItem = function hasItem(itemIdentifier) {
+                return itemIdentifier && self.itemStore.has(itemIdentifier);
+            };
+
+            /**
+             * Check whether we have the next item in the store
+             *
+             * @param {String} itemIdentifier - the CURRENT item identifier
+             * @returns {Boolean}
+             */
+            this.hasNextItem = function hasNextItem(itemIdentifier) {
+                var sibling = navigationHelper.getNextItem(this.getDataHolder().get('testMap'), itemIdentifier);
+
+                return sibling && self.hasItem(sibling.id);
+            };
+
+            /**
+             * Check whether we have the previous item in the store
+             *
+             * @param {String} itemIdentifier - the CURRENT item identifier
+             * @returns {Boolean}
+             */
+            this.hasPreviousItem = function hasPreviousItem(itemIdentifier) {
+                var sibling = navigationHelper.getPreviousItem(this.getDataHolder().get('testMap'), itemIdentifier);
+
+                return sibling && self.hasItem(sibling.id);
+            };
+
+            /**
+             * Offline ? We try to navigate offline, or just say 'ok'
+             *
+             * @param {String} action - the action name (ie. move, skip, timeout)
+             * @param {Object} actionParams - the parameters sent along the action
+             * @returns {Object} action result
+             */
+            this.offlineAction = function offlineAction(action, actionParams) {
+                return new Promise(function(resolve) {
+                    var newTestContext,
+                        result = { success: true },
+                        blockingActions = ['exitTest', 'timeout'],
+                        testData = self.getDataHolder().get('testData'),
+                        testContext = self.getDataHolder().get('testContext'),
+                        testMap = self.getDataHolder().get('testMap');
```

Comment:
Since `getDataHolder()` is called several times, should not it be bette to put its result in a variable instead?